File: algorithms/algorithms-cplusplus/leetcode/Question_0720_Longest_Word_in_Dictionary.cc
#include <string>
#include <vector>

using namespace std;

class TrieNode {
 public:
  TrieNode *child[26];
  bool isWord;
  string word;
  TrieNode() : isWord(false), word("") {
    for (auto &a : child) {
      a = NULL;
    }
  }
};

class Solution {
 public:
  string longestWord(vector<string>& words) {
    TrieNode *root = new TrieNode();
    for (string& w : words) {
      buildTree(w, root);
    }
    string res = "";
    dfs(root, 0, res);
    return res;
  }

 private:
  void buildTree(string& word, TrieNode *root) {
    TrieNode *p = root, *next;
    for (char c : word) {
      if (!p->child[c - 'a']) {
        p->child[c - 'a'] = new TrieNode();
      }
      p = p->child[c - 'a'];
    }
    p->isWord = true;
    p->word = word;
  }

  void dfs(TrieNode *root, int cnt, string& res) {
    cnt++;
    for (TrieNode *p : root->child) {
      if (!p || !p->isWord) {
        continue;
      }
      if (cnt > res.size()) {
        res = p->word;
      }
      dfs(p, cnt, res);
    }
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0720_Longest_Word_in_Dictionary.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0720_Longest_Word_in_Dictionary {
    class TrieNode {
        TrieNode[] children;
        String word;

        TrieNode() {
            children = new TrieNode[26];
        }
    }

    private String res;

    public String longestWord(String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            buildTree(word, root);
        }
        res = "";
        dfs(root, 0);
        return res;
    }

    private void buildTree(String word, TrieNode p) {
        for (char c : word.toCharArray()) {
            if (p.children[c - 'a'] == null) {
                p.children[c - 'a'] = new TrieNode();
            }
            p = p.children[c - 'a'];
        }
        p.word = word;
    }

    private void dfs(TrieNode root, int cnt) {
        cnt++;
        for (TrieNode p : root.children) {
            if (p == null || p.word == null) {
                continue;
            }
            if (cnt > res.length()) {
                res = p.word;
            }
            dfs(p, cnt);
        }
    }
}
